### PR TITLE
Enable customizing MethodAdapterRegistry

### PR DESCRIPTION
**Pull request by bailan**
Make MethodAdapterRegistry customizable, so we can plugin custom argument builders and response builders.

This pull request adds MethodAdapterRegistry to RestLiConfig and updates BaseRestLiServer to use custom MethodAdapterRegistry when provided. MethodAdapterRegistry is also extracted to an interface and existing one is renamed to DefaultMethodAdapterRegistry.

### CODE REVIEW

File: restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java
Comment (ktsimpso) [reply restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java:94]: It looks like `ErrorResponseBuilder` is required for the `batch` methods.

Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java:94]: I'd argue that it does not have to be implemented with `ErrorResponseBuilder`. Anyway, if you think and we can decide to have this dependency, I can add it to the interface with an initialization method like https://github.com/linkedin/rest.li/blob/7ecc85455b8a00e6d71986f97a3c02242749a9b4/restli-server/src/main/java/com/linkedin/restli/server/RestLiDocumentationRequestHandler.java#L30

File: restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java]: Deleted the constructor in 947ed58

File: restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java]: Deleted the Builder in 947ed58

File: restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/BaseRestLiServer.java]: Updated to create `DefaultMethodAdapterRegistry` using global `ErrorResponseBuilder` in `RestLiConfnig` in 947ed58

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:109]: This still has to be null for the get method.
```
  public MethodAdapterRegistry getMethodAdapterRegistry() {
    return Optional.ofNullable(_methodAdapterRegistry)
            .orElse(new DefaultMethodAdapterRegistry(new ErrorResponseBuilder(_errorResponseFormat)));
  }
  ```

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (ktsimpso) [inline restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:617]: I assume we need to default this for backwards compatibility?

Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:617]: Not only for backwards compatibility. If users do not specify `MethodAdapterRegistry` in `RestLiConfig`, this guarantees the use of the default one.

**Comment by bailan** [comment]
> Do we need a major version bump for these changes? (Can we try snapshotting this and testing against some repos?)

We probably need a major version bump.

I did a test by overriding `GetResponseBuilder` and verified the response was built by the overridden builder.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodAdapterRegistry.java
Comment (evanw555) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodAdapterRegistry.java:7]: Simply changing this class to an interface is backward-incompatible. There may be places where this is constructed explicitly. I suggest keeping `MethodAdapterRegistry`, deprecating it (with a javadoc saying to use `DefaultMethodAdapterRegistry`), and having its implementation defer to the `DefaultMethodAdapterRegistry` (perhaps by extending it). This means we will need to introduce the interface as a new file with a unique name.

Since I don't want something like `MethodAdapterRegistryInterface`, I suggest we consider renaming this altogether. I propose using the naming convention `MethodBuildersRegistry`, `DefaultMethodBuildersRegistry`, and the now-deprecated class `MethodAdapterRegistry` will extend `DefaultMethodBuildersRegistry` (so that it relies on its functionality).

The exact naming can be adjusted, but my core idea is to use new naming.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java
Comment (evanw555) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java:66]: This is backward-incompatible. Please deprecate the existing constructors, and have them defer to the new classes while also maintaining existing behavior. Please leave a javadoc explaining the suggested alternative for the user.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java
Comment (evanw555) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/response/RestLiResponseHandler.java:83]: Don't remove this just yet. Leave it deprecated. We can remove all of the deprecated stuff together in a follow-up PR.

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (evanw555) [inline restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:606]: If we follow my advice about altering the naming, the naming here should be consistent. The exact naming is up for debate.

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (evanw555) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:617]: I'm skeptical about including the default-constructed object in this override getter. I'm not sure if this will cause any real issues, but we're constructing a new `ErrorResponseBuilder` here rather than using the one defined in `BaseRestLiServer`. It may be better to return null (or an `Optional`) here, then construct the default in `BaseRestLiServer` using the existing `ErrorResponseBuilder`.

**Comment by evanw555** [comment]
@bailan Do not bump the major version. Please follow my compatibility recommendations and bump the minor version (since we're introducing new interfaces and methods). We can clean up the deprecated stuff in a follow-up PR that will be bundled with the next major version.

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java:617]: I think there are two folds of you comments: (1) using existing `ErrorResponseBuilder` and (2) having default in override getter (RestLiConfig).

For (1), the reason to construct the default in the config is to make sure it uses the `ErrorResponseFormat` in the config rather than the local one embedded in `ErrorResponseBuilder`. This should be the intention based on the following analysis. So, I think this would be a good opportunity to make this idea explicit.

Only three dependency chains I can find for `BaseRestLiServer`. They are end at `RestLiServer`.
```
BaseRestLiServer <- RestRestLiServer <- RestLiServer
BaseRestLiServer <- RestRestLiServer <- StreamRestLiServer <- AttachmentHandlingRestLiServer <- RestLiServer
BaseRestLiServer <- StreamRestLiServer <- AttachmentHandlingRestLiServer <- RestLiServer
```
In `RestLiServer`, the `ErrorResponseBuilder`, `RestRestLiServer` and `AttachmentHandlingRestLiServer` are initialized as follows.
```
    ErrorResponseBuilder errorResponseBuilder = new ErrorResponseBuilder(config.getErrorResponseFormat());

    _restRestLiServer = new RestRestLiServer(config,
        resourceFactory,
        engine,
        rootResources,
        errorResponseBuilder);
    _streamRestLiServer = new AttachmentHandlingRestLiServer(config,
        resourceFactory,
        engine,
        rootResources,
        errorResponseBuilder);
```
https://github.com/linkedin/rest.li/blob/710acd54c1ce4ec8cc6615ccf1d95250e8a09ed2/restli-server/src/main/java/com/linkedin/restli/server/RestLiServer.java#L112

Given that this particular builder is immutable (builder might not be the best name, but this is a different issue), this suggests that constructing it in the config or using the local one is the same. Therefore, I think the intention is to have the `DefaultMethodAdapterRegistry` using the `ErrorResponseFormat` in the config.


For (2), I could move this out of the config, but in the `BaseRestLiServer` the initialization will not use the local `ErrorResponseBuilder`.
```
_methodBuildersRegistry = Optional.ofNullable(config.getMethodAdapterRegistry())
            .orElse(new DefaultMethodAdapterRegistry(new ErrorResponseBuilder(_errorResponseFormat)));
```

File: restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodBuildersRegistry.java
Comment (junchuanwang) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodBuildersRegistry.java:6]: java doc for public class please

File: restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java
Comment (junchuanwang) [inline restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java:61]: newline for curly brace please

File: restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java
Comment (junchuanwang) [inline restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java:69]: else needs to be in separate line

**Review by junchuanwang — COMMENTED** [review]
please follow Evan's suggestion for backward incompatible changes

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (evanw555) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java]: You bring up a good point about "making this idea explicit". The issue is that if we don't use the error response builder in the constructor, then the API is misleading. If the constructor takes in an error response builder, then it implies that this is the one that will be used.

In order to preserve consistency, we should go with one of two routes:
1. Remove the `ErrorResponseBuilder` argument from the constructor and the constructor of all subclasses. The config argument is present, so response builders should just be constructed using the error response format config. Since this is backward-compatible (users may define base classes extending BaseRestLiServer and pass up their own response builder). This is backward-incompatible, so instead please deprecate the existing constructors (with javadoc notes) and create new package-private constructors without the error response builder argument.
2. Follow my original suggestion and use the error response builder provided in the constructor.

I suggest going with option 1, since option 2 would result in a config method that returns an `Optional`, and it would be incompatible to go back.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodBuildersRegistry.java
Comment (evanw555) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodBuildersRegistry.java:7]: After some offline discussion, I've decided that the new preferred naming for this interface is `MethodAdapterProvider`. Please update all subclasses/tests/methods accordingly.

Sorry for the overhead, just trying to get the naming right 😃

Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/methods/MethodBuildersRegistry.java]: Sounds good. Updated in dce626d

File: restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java
Comment (bailan) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java]: Updated

File: restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java
Comment (bailan) [reply restli-int-test/src/test/java/com/linkedin/restli/examples/method/TestCustomMethodBuildersRegistry.java]: Updated

File: restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java
Comment (bailan) [reply restli-server/src/main/java/com/linkedin/restli/server/RestLiConfig.java]: Thank you for the suggestions. I updated code with option 1 dce626d

**Comment by evanw555** [comment]
LGTM, I'm going to run internal wc-tests today and I'll approve/merge once those pass

**Comment by evanw555** [comment]
Tests pass. Going to merge.

**Review by evanw555 — APPROVED** [review]

**Pull request merged**
